File: push/src/main/java/top/trumeet/mipush/provider/db/EventDb.java
package top.trumeet.mipush.provider.db;

import static top.trumeet.mipush.provider.DatabaseUtils.daoSession;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.Nullable;

import com.xiaomi.push.service.MIPushEventProcessor;
import com.xiaomi.xmpush.thrift.XmPushActionContainer;
import com.xiaomi.xmpush.thrift.XmPushActionRegistrationResult;
import com.xiaomi.xmsf.push.utils.RegSecUtils;
import com.xiaomi.xmsf.utils.ConvertUtils;

import org.greenrobot.greendao.query.QueryBuilder;
import org.greenrobot.greendao.query.WhereCondition;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import top.trumeet.common.utils.DatabaseUtils;
import top.trumeet.common.utils.Utils;
import top.trumeet.mipush.provider.event.Event;
import top.trumeet.mipush.provider.event.EventType;
import top.trumeet.mipush.provider.gen.db.EventDao;

/**
 * @author Trumeet
 * @date 2017/12/23
 */

public class EventDb {
    public static final String AUTHORITY = "top.trumeet.mipush.providers.EventProvider";
    public static final String BASE_PATH = "EVENT";
    public static final Uri CONTENT_URI = Uri.parse("content://" + AUTHORITY + "/" + BASE_PATH);

    private static DatabaseUtils getInstance(Context context) {
        return new DatabaseUtils(CONTENT_URI, context.getContentResolver());
    }

    public static class RegistrationInfo {
        public Set<String> registered = new HashSet<>();
        public Set<String> unregistered = new HashSet<>();
    }

    public static long insertEvent(Event event) {
        if (event.getType() == Event.Type.SendMessage) {
            Utils.setLastReceiveTime(event.getPkg(), event.getDate());
        }
        return daoSession.insert(event);
    }

    public static long insertEvent(@Event.ResultType int result,
                                   EventType type) {
        return insertEvent(type.fillEvent(new Event(null
                , type.getPkg()
                , type.getType()
                , Utils.getUTC().getTime()
                , result
                , type.getInfo()
                , null
                , null
                , type.getPayload()
                , Utils.getRegSec(type.getPkg())
        )));
    }

    public static List<Event> queryByPage(
            int pageIndex, int pageSize,
            @Nullable Set<Integer> types,
            @Nullable String pkg, @Nullable String text) {
        return query((pageIndex - 1) * pageSize, pageSize, types, pkg, text);
    }

    public static List<Event> query(
            int skip, int limit,
            @Nullable Set<Integer> types,
            @Nullable String pkg, @Nullable String text) {
        QueryBuilder<Event> query = daoSession.queryBuilder(Event.class)
                .orderDesc(EventDao.Properties.Date)
                .limit(limit)
                .offset(skip);
        if (pkg != null && !pkg.trim().isEmpty()) {
            query.where(EventDao.Properties.Pkg.eq(pkg));
        }
        if (types != null && !types.isEmpty()) {
            query.where(EventDao.Properties.Type.in(types));
        }
        if (text != null && !text.trim().isEmpty()) {
            query.where(EventDao.Properties.Info.like("%" + text + "%"));
        }
        return query.list();
    }

    public static void deleteHistory() {
        String data = (Utils.getUTC().getTime() - 1000L * 3600L * 24 * 7) + "";
        QueryBuilder<Event> query = daoSession.queryBuilder(Event.class)
                .where(EventDao.Properties.Type.in(Event.Type.RECEIVE_PUSH, Event.Type.REGISTER, Event.Type.Command))
                .where(EventDao.Properties.Date.lt(data));
        query.buildDelete().executeDeleteWithoutDetachingEntities();
    }

    public static RegistrationInfo queryRegistered() {
        QueryBuilder<Event> query = daoSession.queryBuilder(Event.class)
                .where(EventDao.Properties.Type.in(Event.Type.RegistrationResult, Event.Type.UnRegistration))
                .where(new WhereCondition.StringCondition("1" +
                        " GROUP BY " + EventDao.Properties.Pkg.columnName +
                        " HAVING MAX(" + EventDao.Properties.Date.columnName + ")"));
        List<Event> events = query.list();

        RegistrationInfo info = new RegistrationInfo();
        for (Event event : events) {
            XmPushActionContainer container = MIPushEventProcessor.buildContainer(event.getPayload());
            XmPushActionRegistrationResult data = null;
            try {
                data = (XmPushActionRegistrationResult)
                        ConvertUtils.getResponseMessageBodyFromContainer(container,
                                RegSecUtils.getRegSec(container));
            } catch (Exception ignored) {
            }
            if (event.getType() == Event.Type.RegistrationResult && (data == null || data.errorCode == 0)) {
                info.registered.add(event.getPkg());
            } else {
                info.unregistered.add(event.getPkg());
            }
        }
        return info;
    }

    public static long getLastReceiveTime(String packageName) {
        Long time = Utils.getLastReceiveTime(packageName);
        if (time != null) {
            return time;
        }

        HashSet<Integer> types = new HashSet<>();
        types.add(Event.Type.SendMessage);
        List<Event> events = EventDb.query(0, 1, types,
                packageName, null);
        long lastReceiveTime = 0;
        if (!events.isEmpty()) {
            lastReceiveTime = events.get(0).getDate();
        }
        Utils.setLastReceiveTime(packageName, lastReceiveTime);
        return lastReceiveTime;
    }

}

File: push/src/main/java/top/trumeet/mipush/provider/db/EventDb.kt
package top.trumeet.mipush.provider.db

import android.content.Context
import android.net.Uri
import androidx.sqlite.db.SimpleSQLiteQuery
import com.nihility.XMPushUtils
import com.xiaomi.xmpush.thrift.XmPushActionRegistrationResult
import com.xiaomi.xmsf.push.utils.RegSecUtils
import com.xiaomi.xmsf.utils.ConvertUtils
import kotlinx.coroutines.runBlocking
import top.trumeet.common.utils.DatabaseUtils
import top.trumeet.common.utils.Utils
import top.trumeet.mipush.provider.DatabaseUtils.eventDao
import top.trumeet.mipush.provider.entities.Event
import top.trumeet.mipush.provider.event.EventType

/**
 * @author Trumeet
 * @date 2017/12/23
 */
object EventDb {
    const val AUTHORITY = "top.trumeet.mipush.providers.EventProvider"
    const val BASE_PATH = "EVENT"

    @JvmField
    val CONTENT_URI: Uri = Uri.parse("content://$AUTHORITY/$BASE_PATH")

    @JvmStatic
    private fun getInstance(context: Context): DatabaseUtils {
        return DatabaseUtils(CONTENT_URI, context.contentResolver)
    }

    class RegistrationInfo {
        @JvmField
        var registered: MutableSet<String> = HashSet()

        @JvmField
        var unregistered: MutableSet<String> = HashSet()
    }

    @JvmStatic
    fun insertEvent(event: Event): Long = runBlocking {
        if (event.type == Event.Type.SendMessage) {
            Utils.setLastReceiveTime(event.pkg, event.date)
        }
        eventDao.insert(event)
    }

    @JvmStatic
    fun insertEvent(@Event.ResultType result: Int, type: EventType): Long {
        return insertEvent(createEvent(result, type))
    }

    @JvmStatic
    fun createEvent(@Event.ResultType result: Int, type: EventType): Event {
        return Event(
            id = null,
            pkg = type.pkg ?: "",
            type = type.type,
            date = Utils.getUTC().time,
            result = result,
            info = type.info,
            payload = type.payload,
            regSec = Utils.getRegSec(type.pkg ?: "")
        )
    }

    @JvmStatic
    fun queryById(
        lastId: Long?,
        size: Int,
        types: Set<Int>?,
        pkg: String?,
        text: String?
    ): List<Event> = runBlocking {
        val queryBuilder = StringBuilder("SELECT * FROM EVENT WHERE 1=1")
        val args = mutableListOf<Any>()
        if (lastId != null) {
            queryBuilder.append(" AND id < ?")
            args.add(lastId)
        }
        if (!pkg.isNullOrBlank()) {
            queryBuilder.append(" AND pkg = ?")
            args.add(pkg)
        }
        if (!types.isNullOrEmpty()) {
            queryBuilder.append(" AND type IN (")
            queryBuilder.append(types.joinToString(",") { "?" })
            queryBuilder.append(")")
            args.addAll(types)
        }
        if (!text.isNullOrBlank()) {
            queryBuilder.append(" AND dev_info LIKE ?")
            args.add("%$text%")
        }
        queryBuilder.append(" ORDER BY id DESC LIMIT ?")
        args.add(size)

        eventDao.queryRaw(SimpleSQLiteQuery(queryBuilder.toString(), args.toTypedArray()))
    }

    @JvmStatic
    fun queryByPage(
        pageIndex: Int,
        pageSize: Int,
        types: Set<Int>?,
        pkg: String?,
        text: String?
    ): List<Event> {
        return query((pageIndex - 1) * pageSize, pageSize, types, pkg, text)
    }

    @JvmStatic
    fun query(
        skip: Int,
        limit: Int,
        types: Set<Int>?,
        pkg: String?,
        text: String?
    ): List<Event> = runBlocking {
        val queryBuilder = StringBuilder("SELECT * FROM EVENT WHERE 1=1")
        val args = mutableListOf<Any>()
        if (!pkg.isNullOrBlank()) {
            queryBuilder.append(" AND pkg = ?")
            args.add(pkg)
        }
        if (!types.isNullOrEmpty()) {
            queryBuilder.append(" AND type IN (")
            queryBuilder.append(types.joinToString(",") { "?" })
            queryBuilder.append(")")
            args.addAll(types)
        }
        if (!text.isNullOrBlank()) {
            queryBuilder.append(" AND dev_info LIKE ?")
            args.add("%$text%")
        }
        queryBuilder.append(" ORDER BY date DESC LIMIT ? OFFSET ?")
        args.add(limit)
        args.add(skip)

        eventDao.queryRaw(SimpleSQLiteQuery(queryBuilder.toString(), args.toTypedArray()))
    }

    @JvmStatic
    fun deleteHistory() = runBlocking {
        val data = Utils.getUTC().time - 1000L * 3600L * 24 * 7
        eventDao.deleteHistory(data)
    }

    @JvmStatic
    fun queryRegistered(): RegistrationInfo = runBlocking {
        val events = eventDao.queryRegisteredStatus()
        val info = RegistrationInfo()
        for (event in events) {
            val container = XMPushUtils.packToContainer(event.payload)
            var data: XmPushActionRegistrationResult? = null
            try {
                data = ConvertUtils.getResponseMessageBodyFromContainer(
                    container,
                    RegSecUtils.getRegSec(container)
                ) as XmPushActionRegistrationResult
            } catch (_: Exception) {
            }
            if (event.type == Event.Type.RegistrationResult && (data == null || data.errorCode.toInt() == 0)) {
                info.registered.add(event.pkg)
            } else {
                info.unregistered.add(event.pkg)
            }
        }
        info
    }

    @JvmStatic
    fun getLastReceiveTime(packageName: String): Long = runBlocking {
        val time = Utils.getLastReceiveTime(packageName)
        if (time != null) {
            return@runBlocking time
        }

        val event = eventDao.getLastEventByType(packageName, Event.Type.SendMessage)
        val lastReceiveTime = event?.date ?: 0L
        Utils.setLastReceiveTime(packageName, lastReceiveTime)
        lastReceiveTime
    }
}
